File: CPP/8_StringToInteger.cpp
// 8 - String to Integer

int myAtoi(string s) {
    long res = 0;
    int idx = 0;

    while(idx < s.size() && s[idx] == ' ') idx++;

    bool isNegative = s[idx] == '-';
    if(s[idx] == '+' || s[idx] == '-') {
        idx++;
    }

    while(idx < s.size() && isdigit(s[idx])) {
        res = res * 10 + (s[idx] - '0');
        if(res > INT_MAX || (isNegative && -res < INT_MIN)) {
            return isNegative ? INT_MIN : INT_MAX;
        }
        idx++;
    }
    return isNegative ? int(-res) : int(res);
}

File: Java/8_StringToInteger.java
// 8 - String To Integer

public int myAtoi(String s) {
    if(s.isEmpty()) return 0;

    long res = 0;
    int idx = 0;

    while(idx < s.length() && s.charAt(idx) == ' ') {
        idx++;
    }

    if(idx >= s.length()) return 0;

    boolean isNegative = s.charAt(idx) == '-';
    if(s.charAt(idx) == '+' || s.charAt(idx) == '-') {
        idx++;
    }

    while(idx < s.length() && s.charAt(idx) >= '0' && s.charAt(idx) <= '9') {
        res = res * 10 + (s.charAt(idx) - '0');
        if(res > Integer.MAX_VALUE || (isNegative && -res < Integer.MIN_VALUE)) {
            return isNegative ?  Integer.MIN_VALUE : Integer.MAX_VALUE;
        }
        idx++;
    }
    
    return isNegative ? (int)(-res) : (int)res;
}